### PR TITLE
3D Improved UNet for segmenting 3D prostate data - Nathan King (47399718)

### PR DESCRIPTION
**Pull request by nk20032003**
Hello Teaching Team,

This is my pull request for the COMP3710 Project Semester 2 2024 to merge my topic-recognition branch to Shakes' topic recognition branch.

I completed the 3D Improved UNet task (Task 4) to segment the 3D prostate MRI dataset (https://data.csiro.au/collection/csiro:51392v2?redirected=true) with all class labels achieving a minimum dice similarity coefficient of 0.7. Regarding the dice similarity coefficients for the test data, I managed to produce 0.9911 for background, 0.9741 for body, 0.8620 for bone, 0.8717 for bladder, 0.7443 for rectum and 0.7318 for prostate. Additionally, the multiclass dice similarity coefficient produced was 0.8625.

All relevant files for this task are contained in the '3D Improved UNet for segmenting 3D prostate data - Nathan King (47399718)' folder. These files are listed below.
- readme_assets: This folder contains visual assets for the README.md file.
- README.md: This file contains documentation of the project, including model explanations, script instructions, training plots and example visualisations.
- modules.py: This script constructs the individual modules of the 3D Improved UNet Model. Additionally, the individual modules are combined to build the entire model.
- dataset.py: This script contains a function used to downscale and load the training, testing and validation data from the MRI and label .nii files.
- train.py: This script executes the training, validating, testing and saving process of the 3D Improved UNet Model. Additionally, accuracy, loss, multiclass dice coefficient and dice similarity coefficient plots are created and saved to visualise the performance of the model.
- predict.py: This script executes the loading and testing process of the trained 3D Improved UNet Model. Additionally, examples showing the input, true segmentation and predicted segmentation are visualised and saved to depict the performance of the model.

Please read the script running instructions to successfully use train.py and predict.py.

Thank you,
Nathan King 47399718

### CODE REVIEW

**Comment by celinsze** [comment]
**This is an initial inspection, no action is required at this point**

Recognition problem: total 20 Marks
Problem Solving: (5)
•	Algorithm solves the problem and achieves a Dice score of above 0.73 for all labels. 
Algorithm implementation: (3)
•	Algorithm implemented in Tensorflow for the improved UNet3D model and functions.
Design: Good coding design (1)
Commenting: Clear and sufficient comments throughout the code (1)
Difficulty: 3D Improved UNet – Hard (10)

**Comment by gayanku** [comment]
## Marking
<table><tbody>
  <tr><td colspan='3'><b>Good/OK/Fair Practice (Design/Commenting, TF/Torch Usage)</b></td> </tr>
	<tr><td></td><td>Good design and implementation. </td><td></td></tr>
	<tr><td></td><td>Spacing and comments. </td><td></td></tr>
	<tr><td></td><td>Header blocks. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Recognition Problem</b></td></tr>
	<tr><td></td><td>Good solution to problem. </td><td></td></tr>
	<tr><td></td><td>Driver Script present. </td><td></td></tr>
	<tr><td></td><td>File structure present. </td><td></td></tr>
	<tr><td></td><td>Good Usage & Demo & Visualisation & Data usage. </td><td></td></tr>
	<tr><td></td><td>Module present. </td><td></td></tr>
	<tr><td></td><td>Commenting present. </td><td></td></tr>
	<tr><td></td><td>No Data leakage found. </td><td></td></tr>
	<tr><td></td><td>Difficulty : Hard. UNet3D (Hard Difficulty)</td><td></td></tr>
	
  <tr><td colspan='3'><b>Commit Log</b></td></tr>
	<tr><td></td><td>Good Meaningful commit messages. </td><td></td></tr>
	<tr><td></td><td>Good Progressive commits. </td><td></td></tr>
	
  <tr><td colspan='3'><b>Documentation</b></td></tr>
	<tr><td></td><td>Readme :Good. </td><td></td></tr>
	<tr><td></td><td>Model/technical explanation :Good. </td><td></td></tr>
	<tr><td></td><td>Description and Comments :Good. </td><td></td></tr>
	<tr><td></td><td>Markdown used and PDF submitted. PDF checked.</td><td></td></tr>
	
  <tr><td colspan='3'><b>Pull Request</b></td></tr>
	<tr><td></td><td>Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch). On time submission.</td><td></td></tr>
	<tr><td></td><td>No Feedback required. </td><td></td></tr>
	<tr><td></td><td>Request Description is good. </td><td></td></tr>
  <tr><td colspan='2'><b>TOTAL</b></td><td>0</td></tr>
</tbody></table>

Marked as per the due date and changes after which aren't necessarily allowed to contribute to grade for fairness.
_Subject to approval from Shakes_

**Comment by shakes76** [comment]
Segmentations not one hot encoded -2 additionally

**Comment by shakes76** [comment]
Correction, segmentations do use one-hot encoding +2
